src/core/execute.c

```diff
@@ -2568,82 +2693,69 @@ static int acquire_credentials(
-                         * we are fine, too. */
-                        log_debug_errno(r, "Couldn't read inherited credential '%s', skipping: %m", lc->path);
-                        continue;
+                if (!path_is_absolute(lc->path))
+                        is_dir = false;
+                else {
+                        sub_fd = open(lc->path, O_DIRECTORY|O_CLOEXEC|O_RDONLY);
+                        if (sub_fd < 0) {
+                                if (errno == ENOTDIR)
+                                        is_dir = false;
+                                else
+                                        return -errno;
+                        }
                 }
-                if (r < 0)
-                        return log_debug_errno(r, "Failed to read credential '%s': %m", lc->path);
 
-                if (lc->encrypted) {
-                        _cleanup_free_ void *plaintext = NULL;
-                        size_t plaintext_size = 0;
+                if (!is_dir) {
+                        r = load_credential(context, params, lc, unit, dfd, uid, ownership_ok, &left);
+                        if (r < 0)
+                                return r;
+
+                } else {
+                        char **path;
+                        _cleanup_strv_free_ char **credpaths = NULL;
+
+                        r = recurse_dir_at(
+                                        dfd,
+                                        lc->path,
+                                        /* statx_mask= */ 0,
+                                        /* n_depth_max= */ UINT_MAX,
+                                        RECURSE_DIR_IGNORE_DOT|RECURSE_DIR_ENSURE_TYPE,
+                                        append_path_recurse_dir_cb,
+                                        &credpaths);
 
-                        r = decrypt_credential_and_warn(lc->id, now(CLOCK_REALTIME), NULL, data, size, &plaintext, &plaintext_size);
                         if (r < 0)
                                 return r;
 
-                        free_and_replace(data, plaintext);
-                        size = plaintext_size;
-                }
+                        STRV_FOREACH(path, credpaths) {
+                                _cleanup_(exec_load_credential_freep) ExecLoadCredential *sub_lc = NULL;
+                                _cleanup_free_ char *cred = NULL, *sub_id = NULL;
 
-                add = strlen(lc->id) + size;
-                if (add > left)
-                        return -E2BIG;
+                                r = path_make_relative(lc->path, *path, &cred);
+                                if (r < 0)
+                                        return r;
+                                sub_id = strreplace(cred, "/", "_");
 
-                r = write_credential(dfd, lc->id, data, size, uid, ownership_ok);
-                if (r < 0)
-                        return r;
+                                sub_lc = new(ExecLoadCredential, 1);
```

Comment:
no need to allocate this, you could just do this:

```c
r = load_credential(context, params, 
        &(ExecLoadCredential) {
                .id = …,
                .path = *path,
                .ecnrypted = lc->encrypted,
        }, …
```

i.e. use C99 compound literals.